Reduce re-rendering on handleSubmit

JeromeDeLeon opened the pull request.

JeromeDeLeon committed:
Reduce re-rendering

bluebill1049 committed:
reduce code

bluebill1049 committed:
Revert "reduce code"

This reverts commit 0040a2efd28ac87a0e6593f627a711167d2301ec.

JeromeDeLeon commented:
I'll push my changes later and summarize of what I have found.

JeromeDeLeon committed:
Fix e2e test

JeromeDeLeon committed:
Fix useFieldArray test

JeromeDeLeon committed:
Merge branch 'enhancement/handleSubmit' of https://github.com/react-hook-form/react-hook-form into enhancement/handleSubmit

JeromeDeLeon commented on cypress/integration/useFieldArray.ts:
```diff
@@ -80,7 +80,7 @@ context('useFieldArray', () => {
     cy.get('#result').contains('{}');
 
-    cy.get('#renderCount').contains('35');
+    cy.get('#renderCount').contains('27');
```
This is one of the gap I'm talking about.

bluebill1049 replied:
strange...

JeromeDeLeon commented on src/controller.tsx:
```diff
@@ -38,7 +38,7 @@ const Controller = <
     mode: { isOnSubmit, isOnBlur, isOnChange },
     reValidateMode: { isReValidateOnBlur, isReValidateOnSubmit },
-    formState: { isSubmitted },
+    isSubmittedRef,
```
I found out about this one though, if keep using the `formState`, for every keystroke, the whole component will re-render.

bluebill1049 replied:
this one is good!

JeromeDeLeon commented on src/controller.tsx:
```diff
@@ -57,6 +57,7 @@ const Controller = <
   const onFocusRef = React.useRef(onFocus);
   const isNotFieldArray = !isNameInFieldArray(fieldArrayNamesRef.current, name);
   rulesRef.current = rules;
+  const isSubmitted = isSubmittedRef.current;
```
The reason I did this because, I found out that we're using `useCallback` and at first, the value of it is `false` but when you hit submit, the value becomes `true` but not in the case of when accessing it inside `shouldValidate` and reading it because we are using `useCallback` once, it memoize the function once and keep the value to `false` even though it changed.

bluebill1049 replied:
👍

JeromeDeLeon commented on src/useForm.ts:
```diff
@@ -130,7 +130,7 @@ export function useForm<
     isDirty: !isProxyEnabled,
     dirtyFields: !isProxyEnabled,
-    isSubmitted: isOnSubmit,
+    isSubmitted: !isProxyEnabled,
```
@bluebill1049 Not sure why you did this, Could you explain? thanks :smile:

bluebill1049 replied:
`submitted` doesn't really care about proxy because that finally statement with the re-render.

JeromeDeLeon commented on src/useForm.ts:
```diff
@@ -1067,7 +1068,14 @@ export function useForm<
-        reRender();
+
+        if (
+          readFormStateRef.current.isSubmitted ||
```
As you can see, I only put the `reRender` method inside because `isSubmittedRef` is being used on the other part of the app hence need to be updated and so for the rest of the `ref`.

JeromeDeLeon reviewed (commented):
I divided the commits to two, the former which I examined and have questions on it and the latter to which I do not understand whats going with `useFieldArray` at all. You could switch it to it and examine it. For the rest, it has minimal impact with only 1 render being saved and I don't know if it's worth pushing, but with `useFieldArray`, a large gap can be seen.

My feelings for it is kinda not wanna push this through because there's a lot for something so simple like skipping the check.

@bluebill1049 probably skip it because it needs to re-render in case there's an error anyway. See [this](https://github.com/react-hook-form/react-hook-form/blob/43af1e0ebac5583024b9f97fb9da2944be0efaad/src/useForm.ts#L1064), there's no `reRender` getting called.

JeromeDeLeon commented:
It happened again! :laughing: connection error on e2e. @bluebill1049

JeromeDeLeon commented:
At first glance, it was so simple with just checking if it was necessary to render or not, but as I proceed with test, I found it complex.

JeromeDeLeon commented:
If this wouldn't be push through, it's fine by me :smile: I'm more than happy learning with this lib. :+1:

bluebill1049 commented:
> I divided the commits to two, the former which I examined and have questions on it and the former to which I do not understand whats going with `useFieldArray` at all. You switch it to it and examined it. For the rest, it has minimal impact with only 1 render being saved and I don't know if it's worth pushing, but with `useFieldArray`, a large gap can be seen.
> 
> My feelings for it is kinda not wanna push this through because there's a lot for something so simple like skipping the check.
> 
> @bluebill1049 probably skip it because it needs to re-render in case there's an error anyway. See [this](https://github.com/react-hook-form/react-hook-form/blob/43af1e0ebac5583024b9f97fb9da2944be0efaad/src/useForm.ts#L1064), there's no `reRender` getting called.

it's intentional because of that final statement. `try` bit

bluebill1049 commented:
> If this wouldn't be push through, it's fine by me 😄 I'm more than happy learning with this lib. 👍

I will take closer look today, i think there are some good stuff in here. good one.

bluebill1049 committed:
Merge branch 'master' into enhancement/handleSubmit

bluebill1049 committed:
revert render after hanleSubmit
- if you take a look the code you can almost merge the two block
- submit need to render error or resolve errors dispite or proxy setting

bluebill1049 committed:
revert automation tests
